File: question/part6_dynamicProgramming/src/542_01matrix.cpp
/*
542. 01 矩阵
给定一个由 0 和 1 组成的矩阵 mat ，请输出一个大小相同的矩阵，其中每一个格子是 mat 中对应位置元素到最近的 0 的距离。
两个相邻元素间的距离为 1 。
*/

#include "../../utils/common.hpp"

/**
 * 两次动态规划求解
 * 
*/
vector<vector<int>> update01Matrix(vector<vector<int>>& mat) {
    if (mat.empty()) return {};
    int m = mat.size(), n = mat[0].size();
    vector<vector<int>> dp(m, vector<int>(n, INT_MAX - 1));  // 1. 定义dp
    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            if (mat[i][j] == 0) {  // 2. 边界条件和迭代公式
                dp[i][j] = 0;
            } else {
                if (j > 0) {
                    dp[i][j] = min(dp[i][j], dp[i][j - 1] + 1);
                }
                if (i > 0) {
                    dp[i][j] = min(dp[i][j], dp[i - 1][j] + 1);
                }
            }
        }
    }

    for (int i = m-1; i > -1; i--) {
        for (int j = n -1; j > -1; j--) {
            if (dp[i][j] != 0) {  // dp[i][j]为0时直接跳过
                if (j < n - 1) {
                    dp[i][j] = min(dp[i][j], dp[i][j + 1] + 1);
                }
                if (i < m - 1) {
                    dp[i][j] = min(dp[i][j], dp[i + 1][j] + 1);
                }
            }
        }
    }
    return dp;
}

int main(){
    vector<vector<int>> matrix = {{0,1,0},{0,1,0},{0,1,0},{0,1,0},{0,1,0}};
    vector<vector<int>> res = update01Matrix(matrix);
    printVector2d(res);

    return 0;
}

File: question/part6_dynamicProgramming/script/542_01matrix.py
"""
542. 01 矩阵
给定一个由 0 和 1 组成的矩阵 mat ，请输出一个大小相同的矩阵，其中每一个格子是 mat 中对应位置元素到最近的 0 的距离。
两个相邻元素间的距离为 1 。
"""


def update01Matrix(mat):
    dp = [[float("inf")] * len(mat[0]) for _ in range(len(mat))]  # dp[i][j]表示该位置（i， j）到最近0的距离
    for i in range(len(mat)):  # 动态规划: 从左上往右下更新dp
        for j in range(len(mat[0])):
            if mat[i][j] == 0:
                dp[i][j] = 0
            else:
                if j > 0:
                    dp[i][j] = min(dp[i][j - 1] + 1, dp[i][j])  # 从上往下更新
                if i > 0:
                    dp[i][j] = min(dp[i - 1][j] + 1, dp[i][j])  # 从左往右更新

    print(dp)
    for i in range(len(mat) - 1, -1, -1):  # 第二次动态规划: 从右下往左上更新dp
        for j in range(len(mat[0]) - 1, -1, -1):
            if mat[i][j] != 0:
                if j < len(mat[0]) - 1:
                    dp[i][j] = min(dp[i][j + 1] + 1, dp[i][j])  # 从下往上更新
                if i < len(mat) - 1:
                    dp[i][j] = min(dp[i + 1][j] + 1, dp[i][j])  # 从右往左更新
    return dp


if __name__ == '__main__':
    matrix = [[0, 0, 0], [0, 1, 0], [1, 1, 1]]
    print(update01Matrix(matrix))
